plutus-scb/src/Plutus/SCB/Core/ContractInstance.hs

```diff
@@ -0,0 +1,484 @@
+import qualified Ledger.AddressMap                                 as AM
+import           Ledger.Constraints.OffChain                       (UnbalancedTx (..))
+import           Ledger.Slot                                       (Slot)
+import           Wallet.API                                        (signWithOwnPublicKey)
+import           Wallet.Effects                                    (ChainIndexEffect, SigningProcessEffect,
+                                                                    WalletEffect, ownPubKey, startWatching, submitTxn,
+                                                                    transactionConfirmed, walletSlot, watchedAddresses)
+
+import           Plutus.SCB.Command                                (saveBalancedTx, saveBalancedTxResult,
+                                                                    saveContractState, sendContractEvent)
+import           Plutus.SCB.Effects.Contract                       (ContractCommand (..), ContractEffect)
+import qualified Plutus.SCB.Effects.Contract                       as Contract
+import           Plutus.SCB.Effects.EventLog                       (EventLogEffect, runCommand, runGlobalQuery)
+import           Plutus.SCB.Effects.UUID                           (UUIDEffect, uuidNextRandom)
+import           Plutus.SCB.Events                                 (ChainEvent (..))
+import           Plutus.SCB.Events.Contract                        (ContractEvent (..), ContractInboxMessage (..),
+                                                                    ContractInstanceId (..), ContractInstanceState (..),
+                                                                    ContractIteration (..), ContractOutboxMessage (..),
+                                                                    ContractRequest (..), ContractResponse (..),
+                                                                    PartiallyDecodedResponse (..), iterationZero)
+import           Plutus.SCB.Query                                  (ContractInboxState (..))
+import qualified Plutus.SCB.Query                                  as Query
+import           Plutus.SCB.Types                                  (SCBError (..), Source (ContractEventSource, NodeEventSource, UserEventSource, WalletEventSource))
+import           Plutus.SCB.Utils                                  (render, tshow)
+
+import qualified Plutus.SCB.Core.Projections                       as Projections
+
+sendContractStateMessages ::
+    forall t effs.
+        ( Member (EventLogEffect (ChainEvent t)) effs
+        , Member (Error SCBError) effs
+        , Member Log effs
+        )
+    => t
+    -> ContractInstanceId
+    -> ContractIteration
+    -> PartiallyDecodedResponse
+    -> Eff effs ()
+sendContractStateMessages t instanceID iteration newState = do
+    logInfo . render $ "Sending messages for contract" <+> pretty instanceID <+> "at iteration" <+> pretty iteration
```

Comment:
Off topic, but I wonder if we want `logInfoDoc = logInfo . render` (and `log*` variants) as utilities. Seems like we'd get the use from them. 🤔

Reply:
And indeed I could imagine that we may want to configure the rendering along with the logging.